src/main/java/org/radarbase/appserver/service/scheduler/MessageSchedulerService.java

```diff
@@ -48,130 +48,129 @@
+        Map<JobDetail, Set<? extends Trigger>> jobDetailSetMap = new HashMap<>();
 
-    JobDetail jobDetail = getJobDetailForMessage(message, getMessageType(message)).getObject();
+        messages.forEach(
+                (Message message) -> {
+                    log.debug("Message = {}", message);
+                    JobDetail jobDetail = getJobDetailForMessage(message, getMessageType(message)).getObject();
 
-    if (jobDetail != null) {
-      log.debug("Job Detail = {}", jobDetail);
-      Trigger trigger = getTriggerForMessage(message, jobDetail).getObject();
+                    log.debug("Job Detail = {}", jobDetail);
+                    Set<Trigger> triggerSet = new HashSet<>();
+                    triggerSet.add(getTriggerForMessage(message, jobDetail).getObject());
+
+                    jobDetailSetMap.putIfAbsent(jobDetail, triggerSet);
+                });
+        log.info("Scheduling {} messages", messages.size());
+        schedulerService.scheduleJobs(jobDetailSetMap);
+    }
+
+    public void updateScheduledMessage(Message message) {
+        String jobKeyString =
+                NAMING_STRATEGY.getJobKeyName(
+                        message.getUser().getSubjectId(), message.getId().toString());
+        JobKey jobKey = new JobKey(jobKeyString);
+        String triggerKeyString =
+                NAMING_STRATEGY.getTriggerName(
+                        message.getUser().getSubjectId(), message.getId().toString());
+        TriggerKey triggerKey = new TriggerKey(triggerKeyString);
+        JobDataMap jobDataMap = new JobDataMap();
+
+        schedulerService.updateScheduledJob(jobKey, triggerKey, jobDataMap, message);
+    }
+
+    public void deleteScheduledMessages(List<Message> messages) {
+        List<JobKey> keys =
+                messages.stream()
+                        .map(
+                                n ->
+                                        new JobKey(
+                                                NAMING_STRATEGY.getJobKeyName(
+                                                        n.getUser().getSubjectId(), n.getId().toString())))
+                        .collect(Collectors.toList());
+        schedulerService.deleteScheduledJobs(keys);
+    }
+
+    public void deleteScheduledMessage(Message message) {
+        JobKey key =
+                new JobKey(
+                        NAMING_STRATEGY.getJobKeyName(
+                                message.getUser().getSubjectId(), message.getId().toString()));
+        schedulerService.deleteScheduledJob(key);
+    }
+
+    public void sendMessage(FcmDownstreamMessage message) throws Exception {
+        fcmSender.send(message);
+    }
 
-      schedulerService.scheduleJob(jobDetail, trigger);
+    public MessageType getMessageType(Message message) {
+        return message.toString().toUpperCase().contains(MessageType.NOTIFICATION.toString()) ? MessageType.NOTIFICATION : MessageType.DATA;
```

Comment:
You can also use -
```
if(message instanceOf Notification) {
return MessageType.NOTIFICATION;
} else if (message instanceOf Data) {
return MessageType.DATA;
} else {
return MessageType.UNKNOWN;
}
```
Then you don't need to explicitly use toString to determine the type.